File: frontend/src/room/pageKind.ts
/**
 * A KEPT PAGE READS LIKE WHAT IT IS (W4.1, 2026-09-23).
 *
 * The owner, of the PAGES list in the room's sidebar — Estate Dashboard, Store
 * Dashboard, Estate Week, AJI BARN Reorder: *"next thing we need to do is make
 * how each page style work idealy"*. Until this every kept page was a header
 * plus N independent boards, each one packed into the same two-column grid, so
 * a page of four single numbers and a page of five reorder lists came out
 * looking like the same page.
 *
 * WHAT A KIND MAY AND MAY NOT DO. A kind decides how the page is DRAWN. It
 * never changes what is on the page, what any figure says, or the ORDER the
 * person put the analyses in — `groupsFor` below is the whole of the grouping
 * and it is a partition of the list in place, never a sort. Nothing here
 * computes, re-computes or re-words a figure: every number on a kept page is
 * still its pin's own run, drawn by the room's own `Board` (rule 9).
 *
 * WHAT IS DECLARED WHERE, honestly stated (rule 3). The kinds, their meaning
 * and how a page's kind is DERIVED live in `definitions/metrics.yaml`
 * (`pages.kinds`) and reach the room on the page shape — `kind`, `kind_set_by`
 * and, when the server serves them, `kind_options` with the definitions' own
 * words. What lives HERE is the drawing: which analyses group, and the words
 * the control falls back to when a server older than this card serves none.
 * The drawing itself is CSS scoped by `data-page-kind` in `room.css`.
 */
import type { CompositionBlock } from '../types/bob';
import type { Page, PageKind, PageKindOption } from '../types/pins';

/** The contract's four, in the order the control offers them. */
export const PAGE_KINDS: PageKind[] = ['dashboard', 'week', 'list', 'collection'];

/**
 * THE FALLBACK WORDS, in the room's own language rather than the enum's — used
 * only when the server serves no `kind_options`. A raw "collection" in a
 * control would be the schema talking to the owner.
 */
export const KIND_WORDS: Record<PageKind, { label: string; says: string }> = {
  dashboard: { label: 'Numbers you check', says: 'The figures sit together at the top, each with when it was read.' },
  week: { label: 'How it went', says: 'One column, read top to bottom over the page’s dates.' },
  list: { label: 'A working list', says: 'The rows are the point: they take the width, and charts beside them stay small.' },
  collection: { label: 'Saved answers', says: 'A set of kept answers, each drawn as it was kept.' },
};

/** Is this one of the four? A name the room does not know is not a kind. */
export function isPageKind(value: unknown): value is PageKind {
  return typeof value === 'string' && (PAGE_KINDS as string[]).includes(value);
}

/**
 * THE PAGE'S KIND, never null. A response from before this card, a page still
 * loading, and Ungrouped (which is not a page) all read `collection` — which
 * is exactly what a kept page has always drawn as, so nothing breaks on an
 * older response.
 */
export function kindOf(page: Page | null | undefined): PageKind {
  return isPageKind(page?.kind) ? page.kind : 'collection';
}

/** The kinds to offer, in the definitions' words where the server sent them. */
export function kindOptions(page: Page | null | undefined): PageKindOption[] {
  const served = (page?.kind_options ?? []).filter((o) => isPageKind(o.value));
  if (served.length) return served;
  return PAGE_KINDS.map((value) => ({ value, label: KIND_WORDS[value].label,
                                      says: KIND_WORDS[value].says }));
}

/** The words for one kind, the server's where it sent them. */
export function wordsFor(page: Page | null | undefined, kind: PageKind): PageKindOption {
  return kindOptions(page).find((o) => o.value === kind)
    ?? { value: kind, label: KIND_WORDS[kind].label, says: KIND_WORDS[kind].says };
}

/* ------------------------------------------------------------------ shapes */

/**
 * WHAT AN ANALYSIS IS, BY WHAT IT DREW — never by a label a model inferred
 * (rule 9/10). The shape is read off the marks the pin's own run returned,
 * which the room is about to draw anyway.
 *
 *   stat   one number, from one row — a figure or a gauge
 *   rows   the rows themselves are the point — a table or a list
 *   chart  everything the catalogue draws as a shape to read
 *   mixed  more than one family in the same analysis
 *   none   nothing came back to draw (reading, failed, refused, empty)
 */
export type AnalysisShape = 'stat' | 'rows' | 'chart' | 'mixed' | 'none';

/**
 * The mark catalogue's families (metrics.yaml `composition.grammar.marks`).
 * A mark the room has not met is a chart: it is a shape to read, which is what
 * fourteen of the eighteen are.
 */
const STAT = new Set(['figure', 'gauge']);
const ROWS = new Set(['table', 'list']);

function familyOf(kind: string | null | undefined): 'stat' | 'rows' | 'chart' {
  if (STAT.has(kind ?? '')) return 'stat';
  if (ROWS.has(kind ?? '')) return 'rows';
  return 'chart';
}

export function shapeOf(blocks: CompositionBlock[] | null | undefined): AnalysisShape {
  const families = new Set((blocks ?? []).filter((b) => !b.default).map((b) => familyOf(b.kind)));
  if (families.size === 0) return 'none';
  if (families.size > 1) return 'mixed';
  const [only] = [...families];
  // ONE FIGURE IS A STAT TILE; TWO ARE NOT. A dashboard tile is a caption over
  // a number — an analysis that drew three of them is a section, not a tile.
  if (only === 'stat' && (blocks ?? []).filter((b) => !b.default).length > 1) return 'chart';
  return only;
}

/** True for the analyses a dashboard draws as tiles: a single figure, drawn. */
export function isStatTile(shape: AnalysisShape): boolean {
  return shape === 'stat';
}

/* ---------------------------------------------------------------- grouping */

/**
 * HOW THE PAGE IS BROKEN INTO ROWS. A `row` group is a RUN of consecutive
 * stat-tile analyses drawn side by side; every other group holds exactly one
 * analysis, at the full width, where it stands.
 *
 * THE ORDER NEVER MOVES. The groups are `list` cut at the boundaries between
 * tiles and everything else — flattening them returns `list`, which
 * `pageKind.test.ts` holds for every kind.
 */
export interface PinGroup<T> {
  row: boolean;
  items: T[];
  /** Where in the page's own order this group starts — the key for the row. */
  at: number;
}

export function groupsFor<T>(kind: PageKind, items: T[],
                             shapeAt: (item: T, index: number) => AnalysisShape): PinGroup<T>[] {
  const out: PinGroup<T>[] = [];
  // ONLY A DASHBOARD PUTS TWO ANALYSES ON ONE LINE. A week is one column by
  // definition, a list gives its rows the width, and a collection is what a
  // kept page has always been.
  const groups = kind === 'dashboard';
  for (let i = 0; i < items.length; i += 1) {
    if (groups && isStatTile(shapeAt(items[i], i))) {
      const run: T[] = [];
      const at = i;
      while (i < items.length && isStatTile(shapeAt(items[i], i))) { run.push(items[i]); i += 1; }
      i -= 1;
      // A RUN OF ONE IS STILL A TILE. The owner's Store Dashboard opens with a
      // single number above a chart; drawn as a section it would be a 940px
      // box around one figure.
      out.push({ row: true, items: run, at });
      continue;
    }
    out.push({ row: false, items: [items[i]], at: i });
  }
  return out;
}

/**
 * WHICH BLOCK GOES FIRST INSIDE ONE ANALYSIS, on a `week` — "comparisons
 * before lists". A stable partition of ONE analysis's own blocks: the shapes
 * you read come before the rows you scan, and two blocks of the same family
 * keep the order the composer gave them. It moves no analysis, and it is the
 * only place a kind touches anything below the page.
 */
export function blockOrderFor(kind: PageKind, blocks: CompositionBlock[]): CompositionBlock[] {
  if (kind !== 'week' || blocks.length < 2) return blocks;
  const rows = blocks.filter((b) => familyOf(b.kind) === 'rows');
  if (!rows.length || rows.length === blocks.length) return blocks;
  return [...blocks.filter((b) => familyOf(b.kind) !== 'rows'), ...rows];
}

File: backend/app/services/page_kind.py
"""
What KIND of page this is — and therefore how it is DRAWN.

WHAT IT IS FOR (W4.1, 2026-09-23). The owner, of the pages in the room's
sidebar — Estate Dashboard, Store Dashboard, Estate Week, AJI BARN Reorder:
*"next thing we need to do is make how each page style work idealy"*. Four
kept pages doing four different jobs, drawn one way: a header and N boards
packed into the same grid. A page of numbers you check every morning and a
page of rows you work down are not the same page and should not read the same.

A KIND IS PRESENTATION AND NOTHING ELSE. It never changes what is on the page,
what any figure says, which reads run, or the order the person put the
analyses in. Nothing in this module computes, re-computes, rounds or re-words
a figure: it reads stored CALLS — a tool name and its arguments — and returns
one of four words. It runs no read, touches no database, and calls no model.

WHERE THE RULES LIVE. All of them in `pages.kinds` (definitions/metrics.yaml),
read at runtime: the four kinds, what each means, how each draws, and the
ORDERED matchers that derive one. This file holds the mechanism and not one
rule — a matcher naming a condition this file does not implement is a fault,
raised, never a matcher that quietly never fires.

NOTHING HERE LOOKS AT A WORD. Not the page's title, not its purpose, not a
pin's title, not anything a model wrote. A page called "Dashboard" full of
long lists is a `list`.

HOW A PAGE GETS ONE. The owner sets it by hand, or Bob sets it when he builds
or edits the page, and then `pages.kind` holds it and `pages.kind_set_by` says
who. NULL means nobody has said, and the kind is DERIVED on every read — never
stored — so a page whose analyses change is drawn as what it now is.
"""

from __future__ import annotations

import sys
from pathlib import Path
from typing import Any, Iterable, Mapping, Optional, Sequence

_ROOT = Path(__file__).resolve().parents[3]
if str(_ROOT) not in sys.path:
    sys.path.insert(0, str(_ROOT))

from tools._common import load_defs, req  # noqa: E402

# Who said what the kind is. 'user' and 'bob' are stored on the row; 'derived'
# is never stored — it is what a NULL column reads as.
SET_BY_USER = "user"
SET_BY_BOB = "bob"
SET_BY_DERIVED = "derived"
STORED_SET_BY = (SET_BY_USER, SET_BY_BOB)

# The shapes a stored call can be projected to without running it. Anything
# else is unknown, and unknown counts neither for nor against any kind.
SHAPE_LIST = "list"
SHAPE_FIGURE = "figure"
SHAPE_UNKNOWN = "unknown"

# Every condition a matcher may name. A matcher naming anything else raises,
# so the yaml and this file cannot drift apart in silence.
_CONDITIONS = frozenset({
    "min_analyses",
    "min_share_of_figure_analyses",
    "min_share_of_list_analyses",
    "min_distinct_tools",
    "no_read_is_compared",
    "every_read_is_compared",
    "one_window_across_the_page",
    "one_tool_across_the_page",
    "one_grouping_across_the_page",
})
# Keys on a matcher that are prose, not conditions.
_MATCHER_PROSE = frozenset({"name", "gives", "why"})


class KindRefused(ValueError):
    """A kind that is not one of the four. Words a person can act on."""


# ---------------------------------------------------------------------------
# The definitions
# ---------------------------------------------------------------------------

def spec(defs: Optional[Mapping[str, Any]] = None) -> Mapping[str, Any]:
    return req(defs or load_defs(), "pages.kinds")


def kinds(defs: Optional[Mapping[str, Any]] = None) -> list[str]:
    """The four kinds, in the yaml's order. The closed set the API accepts."""
    return [str(k) for k in req(spec(defs), "catalogue")]


def default_kind(defs: Optional[Mapping[str, Any]] = None) -> str:
    return str(req(spec(defs), "default"))


def means(kind: str, defs: Optional[Mapping[str, Any]] = None) -> str:
    """One line, in a person's words, of what this kind of page is."""
    return str(req(spec(defs), f"catalogue.{kind}.means"))


def draws(kind: str, defs: Optional[Mapping[str, Any]] = None) -> dict[str, Any]:
    """
    HOW THIS KIND IS DRAWN, verbatim from the yaml: the width each block shape
    takes, whether a run of consecutive single-figure analyses groups into one
    row, whether the analysis title reads as a head or a caption, and whether
    the page carries a dateline.

    Served to the room on the page itself, the way the window's options are,
    so no component holds a copy of a layout rule.
    """
    return dict(req(spec(defs), f"catalogue.{kind}.draws"))


def options(defs: Optional[Mapping[str, Any]] = None) -> list[dict[str, str]]:
    """
    THE FOUR KINDS AS A PERSON READS THEM — what the owner's own control
    offers. `label` is the kind's `means` and `says` is its `when`, both the
    yaml's words: a control that spelled out "collection" would be the schema
    talking to the owner, and a control holding its own wording would be a
    second copy of a definition (CLAUDE.md rule 3).
    """
    out: list[dict[str, str]] = []
    for kind in kinds(defs):
        label = means(kind, defs).strip()
        out.append({
            "value": kind,
            "label": label[:1].upper() + label[1:],
            "says": " ".join(str(req(spec(defs), f"catalogue.{kind}.when")).split()),
        })
    return out


def check(kind: Any, defs: Optional[Mapping[str, Any]] = None) -> str:
    """One of the four, or refused naming them. Never None: a page has a kind."""
    names = kinds(defs)
    if not isinstance(kind, str) or kind not in names:
        raise KindRefused(
            f"{kind!r} is not a kind of page. One of: {', '.join(names)} — "
            f"or leave it unset and it is worked out from what is on the page."
        )
    return kind


# ---------------------------------------------------------------------------
# What a stored call will draw as
# ---------------------------------------------------------------------------

def _grouping(arguments: Mapping[str, Any]) -> tuple[str, ...]:
    """
    A call's grouping as a tuple, whichever way it was written: absent, a bare
    string, or a list. `()` is the whole scope as one row.
    """
    value = arguments.get("group_by")
    if value is None:
        return ()
    if isinstance(value, str):
        return (value,)
    if isinstance(value, (list, tuple)):
        return tuple(str(v) for v in value)
    return (str(value),)


def _compared(arguments: Mapping[str, Any]) -> bool:
    """Whether the call asked the tool for a baseline to compare against."""
    return bool(arguments.get("compare_to"))


def call_shape(call: Mapping[str, Any], defs: Optional[Mapping[str, Any]] = None) -> str:
    """
    What this stored call will produce, projected from the call alone.

    Deliberately narrow. Two projections are honest without rows — a bounded
    ranked set of named rows is a list, and a measured read with no grouping
    returns one row, which is a figure (agent/default_composition.shape_for
    decides the same way from the rows themselves). Everything else is
    `unknown`, which no matcher counts.
    """
    shapes = req(spec(defs), "derivation.shapes")
    tool = str(call.get("tool") or "")
    arguments = call.get("arguments") if isinstance(call.get("arguments"), Mapping) else {}

    listed = shapes.get(SHAPE_LIST) or {}
    if tool in (listed.get("tools") or []):
        return SHAPE_LIST
    if any(arguments.get(a) is not None for a in (listed.get("arguments") or [])):
        return SHAPE_LIST

    figure = shapes.get(SHAPE_FIGURE) or {}
    if tool in (figure.get("tools") or []):
        if not figure.get("grouping_empty") or not _grouping(arguments):
            return SHAPE_FIGURE
    return SHAPE_UNKNOWN


def analysis_shape(calls: Sequence[Mapping[str, Any]],
                   defs: Optional[Mapping[str, Any]] = None) -> str:
    """
    One analysis's shape: what EVERY call in it projects to, or unknown. One
    call nobody can place makes the analysis unplaced — never the majority of
    its calls, because an analysis is drawn as a whole.
    """
    if not calls:
        return SHAPE_UNKNOWN
    found = {call_shape(c, defs) for c in calls}
    return found.pop() if len(found) == 1 else SHAPE_UNKNOWN


# ---------------------------------------------------------------------------
# The derivation
# ---------------------------------------------------------------------------

class _Page:
    """Everything the matchers may see about a page. Calls, never words."""

    def __init__(self, analyses: Sequence[Sequence[Mapping[str, Any]]],
                 *, has_date_window: bool, defs: Mapping[str, Any]) -> None:
        self.count = len(analyses)
        self.shapes = [analysis_shape(list(a), defs) for a in analyses]
        self.calls = [c for a in analyses for c in a if isinstance(c, Mapping)]
        self.tools = {str(c.get("tool") or "") for c in self.calls}
        self.has_date_window = has_date_window
        self._args = [c.get("arguments") if isinstance(c.get("arguments"), Mapping) else {}
                      for c in self.calls]
        self.groupings = {_grouping(a) for a in self._args}
        self.ranges = {str(a.get("date_range")) for a in self._args if a.get("date_range")}
        self.reads_with_a_range = sum(1 for a in self._args if a.get("date_range"))
        self.compared = [_compared(a) for a in self._args]

    def share(self, shape: str) -> float:
        return (sum(1 for s in self.shapes if s == shape) / self.count) if self.count else 0.0

    def one_window(self) -> bool:
        """
        One window over the whole page: either every read names the same date
        range, or the page itself carries a date window, which re-runs every
        analysis over it whatever each was kept with.
        """
        if self.has_date_window:
            return True
        return bool(self.calls) and self.reads_with_a_range == len(self.calls) and len(self.ranges) == 1


def _holds(condition: str, value: Any, page: _Page) -> bool:
    if condition == "min_analyses":
        return page.count >= int(value)
    if condition == "min_share_of_figure_analyses":
        return page.share(SHAPE_FIGURE) >= float(value)
    if condition == "min_share_of_list_analyses":
        return page.share(SHAPE_LIST) >= float(value)
    if condition == "min_distinct_tools":
        return len(page.tools) >= int(value)
    if condition == "no_read_is_compared":
        return bool(value) is not any(page.compared)
    if condition == "every_read_is_compared":
        return bool(value) is (bool(page.compared) and all(page.compared))
    if condition == "one_window_across_the_page":
        return bool(value) is page.one_window()
    if condition == "one_tool_across_the_page":
        return bool(value) is (len(page.tools) == 1)
    if condition == "one_grouping_across_the_page":
        return bool(value) is (len(page.groupings) == 1)
    raise KeyError(condition)   # unreachable: _conditions_of checked it


def _conditions_of(matcher: Mapping[str, Any]) -> dict[str, Any]:
    conditions = {k: v for k, v in matcher.items() if k not in _MATCHER_PROSE}
    unknown = sorted(set(conditions) - _CONDITIONS)
    if unknown:
        raise KeyError(
            f"pages.kinds.derivation matcher {matcher.get('name')!r} names "
            f"{', '.join(unknown)}, which app/services/page_kind.py does not "
            f"implement. Implement it or take it out — a matcher that cannot "
            f"be read is one that never fires."
        )
    return conditions


def derive(analyses: Sequence[Sequence[Mapping[str, Any]]], *,
           has_date_window: bool = False,
           defs: Optional[Mapping[str, Any]] = None) -> str:
    """
    The kind a page of these analyses is, from what its pins CARRY.

    `analyses` is the page's pins IN THE PAGE'S ORDER, each one its stored
    calls ({tool, arguments}). Pure and deterministic: the same page always
    derives the same kind, and the kind is never written down.

    The first matcher whose every condition holds gives the kind; a page under
    `min_analyses`, or one nothing matches, is the default.
    """
    defs = defs or load_defs()
    rules = req(spec(defs), "derivation")
    page = _Page(analyses, has_date_window=has_date_window, defs=defs)
    if page.count < int(req(rules, "min_analyses")):
        return default_kind(defs)
    for matcher in req(rules, "matchers"):
        conditions = _conditions_of(matcher)
        if all(_holds(k, v, page) for k, v in conditions.items()):
            return check(matcher.get("gives"), defs)
    return default_kind(defs)


def which(matcher_name: str, defs: Optional[Mapping[str, Any]] = None) -> Mapping[str, Any]:
    """The named matcher, for a test or a receipt. Raises if there is none."""
    for matcher in req(spec(defs), "derivation.matchers"):
        if str(matcher.get("name")) == matcher_name:
            return matcher
    raise KeyError(matcher_name)


# ---------------------------------------------------------------------------
# A page's kind, as every surface reports it
# ---------------------------------------------------------------------------

def resolve(stored: Any, stored_set_by: Any,
            analyses: Sequence[Sequence[Mapping[str, Any]]], *,
            has_date_window: bool = False,
            defs: Optional[Mapping[str, Any]] = None) -> tuple[str, str]:
    """
    (kind, kind_set_by) for one page. NEVER (None, ...): a page always has a
    kind, because a page always draws as something.

    A stored kind is the person's or Bob's and is returned as it stands — it
    is not second-guessed against the pins, because somebody said. A stored
    kind that is no longer one of the four (an older name, a hand-edited row)
    is treated as nobody having said, rather than refusing to draw the page.
    """
    defs = defs or load_defs()
    if isinstance(stored, str):
        try:
            return check(stored, defs), (str(stored_set_by) if stored_set_by in STORED_SET_BY
                                         else SET_BY_USER)
        except KindRefused:
            pass
    return derive(analyses, has_date_window=has_date_window, defs=defs), SET_BY_DERIVED


def calls_of(pins: Iterable[Any]) -> list[list[dict]]:
    """
    A page's pins as the derivation takes them: each pin's stored calls, in
    the page's order. Reads `tool_calls` and nothing else — never a title.
    """
    out: list[list[dict]] = []
    for pin in pins:
        calls = getattr(pin, "tool_calls", None)
        if calls is None and isinstance(pin, Mapping):
            calls = pin.get("tool_calls") or pin.get("calls")
        out.append([dict(c) for c in (calls or []) if isinstance(c, Mapping)])
    return out
